File: rb/test/generator_entity_test.rb
# Generator entity test

require "minitest/autorun"
require "json"
require_relative "../DeveloperToolbox_sdk"
require_relative "runner"

class GeneratorEntityTest < Minitest::Test
  def test_create_instance
    testsdk = DeveloperToolboxSDK.test(nil, nil)
    ent = testsdk.Generator(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = generator_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "generator." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    generator_ref01_ent = client.Generator(nil)
    generator_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.generator"), "generator_ref01"))

    generator_ref01_data_result, err = generator_ref01_ent.create(generator_ref01_data, nil)
    assert_nil err
    generator_ref01_data = Helpers.to_map(generator_ref01_data_result)
    assert !generator_ref01_data.nil?

    # LIST
    generator_ref01_match = {}

    generator_ref01_list_result, err = generator_ref01_ent.list(generator_ref01_match, nil)
    assert_nil err
    assert generator_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(generator_ref01_list_result),
      { "id" => generator_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # LOAD
    generator_ref01_match_dt0 = {}
    generator_ref01_data_dt0_loaded, err = generator_ref01_ent.load(generator_ref01_match_dt0, nil)
    assert_nil err
    assert !generator_ref01_data_dt0_loaded.nil?

  end
end

def generator_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "generator", "GeneratorTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = DeveloperToolboxSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["generator01", "generator02", "generator03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID" => idmap,
    "DEVELOPERTOOLBOX_TEST_LIVE" => "FALSE",
    "DEVELOPERTOOLBOX_TEST_EXPLAIN" => "FALSE",
    "DEVELOPERTOOLBOX_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["DEVELOPERTOOLBOX_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["DEVELOPERTOOLBOX_APIKEY"],
      },
      extra || {},
    ])
    client = DeveloperToolboxSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["DEVELOPERTOOLBOX_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["DEVELOPERTOOLBOX_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/generator_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/developer-toolbox-sdk/go"
	"github.com/voxgig-sdk/developer-toolbox-sdk/go/core"

	vs "github.com/voxgig-sdk/developer-toolbox-sdk/go/utility/struct"
)

func TestGeneratorEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Generator(nil)
		if ent == nil {
			t.Fatal("expected non-nil GeneratorEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := generatorBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "generator." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		generatorRef01Ent := client.Generator(nil)
		generatorRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "generator"}, setup.data), "generator_ref01"))

		generatorRef01DataResult, err := generatorRef01Ent.Create(generatorRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		generatorRef01Data = core.ToMapAny(generatorRef01DataResult)
		if generatorRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		generatorRef01Match := map[string]any{}

		generatorRef01ListResult, err := generatorRef01Ent.List(generatorRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		generatorRef01List, generatorRef01ListOk := generatorRef01ListResult.([]any)
		if !generatorRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", generatorRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(generatorRef01List), map[string]any{"id": generatorRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// LOAD
		generatorRef01MatchDt0 := map[string]any{}
		generatorRef01DataDt0Loaded, err := generatorRef01Ent.Load(generatorRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if generatorRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func generatorBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "generator", "GeneratorTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read generator test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse generator test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"generator01", "generator02", "generator03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID": idmap,
		"DEVELOPERTOOLBOX_TEST_LIVE":      "FALSE",
		"DEVELOPERTOOLBOX_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["DEVELOPERTOOLBOX_TEST_GENERATOR_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["DEVELOPERTOOLBOX_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewDeveloperToolboxSDK(core.ToMapAny(mergedOpts))
	}

	live := env["DEVELOPERTOOLBOX_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["DEVELOPERTOOLBOX_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
